src/bigstring_unix_stubs.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2019, Christopher Zimmermann
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
+ * SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+ * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+
+#include <unistd.h>
+#include <caml/mlvalues.h>
+#include <caml/bigarray.h>
+#include <caml/unixsupport.h>
+#include <caml/threads.h>
+
+CAMLprim value
+bigstring_read(value vfd, value vba, value voff, value vlen)
+{
+    void *iobuf = ((char *)Caml_ba_data_val(vba)) + Unsigned_long_val(voff);
+#ifdef Handle_val
+    unsigned len;
+    int err;
+
+    if (Descr_kind_val(fd) == KIND_SOCKET) {
+      SOCKET s = Socket_val(fd);
+      caml_release_runtime_system();
+      if ((err = recv(s, iobuf, Unsigned_int_val(vlen), 0)) < 0)
+        err = WSAGetLastError();
+      else {
+        len = err;
+        err = 0;
+      }
+      caml_acquire_runtime_system();
+    } else {
+      HANDLE h = Handle_val(fd);
+      caml_release_runtime_system();
+      if (ReadFile(h, iobuf, Unsigned_int_val(vlen), &len, NULL))
+        err = 0;
+      else {
+check_error:
+        switch (err = GetLastError()) {
+          case ERROR_BROKEN_PIPE:
+            /* This is no error, but just a closed pipe. */
+            err = len = 0;
+            break;
+          case ERROR_MORE_DATA:
+#if 0
```

Comment:
why is that there? :)

Reply:
The ``ERROR_MORE_DATA``?
That's one thing where Windows behaves different then Unix for message based sockets. Unix discards the rest of the message when the provided buffer is too small, windows returns this error and will return the rest of the message in the next call. So the ``#if 0`` part would mimic Unix behaviour and discard the rest of the message. I'm not sure what to do on that error condition.